Mediasoup update

mcottontensor opened the pull request:
## Relevant components:
- [ ] Signalling server
- [ ] Common library
- [ ] Frontend library
- [ ] Frontend UI library
- [ ] Matchmaker
- [ ] Platform scripts
- [x] SFU

## Problem statement:
We were previously distributing the binaries for the mediasoup workers ourselves but latest versions of mediasoup now provide the binaries themselves.

## Solution
Changed the dependency to mediasoup proper and updated a few pieces of code where the library had changed behaviour.
Have also updated the mediasoup-sdp-bridge package so that it's no longer a prebuilt package and is now built from source that is forked from the original source. This allows history tracking and easier development in the future.

ibc committed:
First commit (WIP)

ibc committed:
README.md: fix typo

ibc committed:
Fix another typo

ibc committed:
Fix typo

ibc committed:
Do not pass router and transportOptions but directly a transport

ibc committed:
Remove sdpSendEndpoint.on('negotiationneeded')

ibc committed:
Improve inline doc

ibc committed:
Remove 'newproducer' and 'newconsumer' events

ibc committed:
Fix API

ibc committed:
Now really fix API

ibc committed:
Address feedback

ibc committed:
more

ibc committed:
cosmetic

ibc committed:
fix typo

ibc committed:
foo

ibc committed:
foo

ibc committed:
Update API

j1elo committed:
[WIP] Implement initially proposed API

This is a proposal implementation of sdp-bridge. It is a work-in-
progress, not thoroughly tested. The focus is on functionality, leaving
architecture considerations aside. Specifically:

* Invalid scenarios are just caught and the process ends with a "BUG"
message.
* Code is currently written for the WebRtcTransport, not for generic
Transports.
* A static RtpCapabilities object (from `BrowserRtpCapabilities`) is
used to represent the RTP caps of remote browsers. This should be worked
on in the future, to obtain these caps from the SDP Answer that a
browser send in response to an SDP Offer. Read more below.
* Some verbose logs are just commented out, instead of using more
adequate technique of filtering by logging level.
* "mediasoup" and "mediasoup-client" are added as dependencies in
package.json.

An important limitation we've found while writing this code is that
there is a "logic loop" in the mediasoup API, which makes it impossible
to create a proper SDP endpoint. This is a summary of the issue:

1. In order to send an SDP Offer, the offerer needs to include its own
SSRC for the medias that are going to be sent.
2. In mediasoup, the SSRC is obtained from the
`Consumer.rtpParameters.encodings[].ssrc`, so a Consumer must have been
created beforehand.
3. In order to create a Consumer, you need to pass its constructor an
`RtpCapabilities` that describes the capabilities of the remote peer.
4. The `RtpCapabilities` that describe what a browser is capable of,
would be obtained from an SDP Answer sent by it.
5. To make a browser send an SDP Answer, you have to first send it an
SDP Offer. GOTO 1.

We're breaking this "loop" by hardcoding in the source code an
RtpCapabilities that represents a typical SDP Answer from the browser,
with 1 audio and 1 video stream. There are capabilities for Firefox,
Safari, and Chrome, of which currently the Chrome ones are in use
because they seem to be compatible with all other browsers, Firefox
included.

Apart from that, another addition to the API was the `addConsumer`
method. Just like sdp-bridge is able to generate an array `Producer[]`
from an SDP Offer, it should also be able to generate an SDP Offer from
an array `Consumer[]`, so it was necessary to add a method that allows
providing these Consumers.

All in all, this only intends to be a preliminary implementation that
opens up further design decisions and conversation.

j1elo committed:
Add Git & Prettier config

j1elo committed:
"mediasoup" dependency set only in peerDependencies (tested with 3.6)

j1elo committed:
[WIP] Fix last bugs

It is now working; verified with openvidu-mediasoup-controller.

j1elo committed:
tsconfig review

* Extend config for Node 10 (official TSConfig Bases).
* Remove settings that were enabled anyway by default.
* Add "removeComments" and "sourceMap".
* Sort alphabetically and format (default settings).
* Review and fix all compilation errors that were caused by
'strictNullChecks = true', caused implicitly by 'strict = true'.

j1elo committed:
tsconfig review

j1elo committed:
SdpUtils: add parsing of "rids" from simulcast SDP

j1elo committed:
Log creation of Producers

j1elo committed:
package.json: add targets build:watch, clean, rebuild, format

j1elo committed:
package.json: move '@types/*' to devDependencies

j1elo committed:
Add lib/ so it's possible to depend on this project by Git URL

j1elo committed:
package.json: update to mediasoup 3.7

Also sort dependencies alphabetically

j1elo committed:
delete lib/ and use a prepare script instead

If a package being installed through git contains a prepare script, its
dependencies and devDependencies will be installed, and the prepare
script will be run, before the package is packaged and installed.

https://docs.npmjs.com/cli/v7/using-npm/scripts#life-cycle-scripts

j1elo committed:
rename 'main.ts' to more standard 'index.ts'

j1elo committed:
add tool "browser-rtpcapabilities"

This tool makes it trivially easy to obtain the RtpCapabilities object
that represents what encodings and other WebRTC features are available
on the web browser the tool runs on.

j1elo committed:
update RtpCapabilities from Firefox 88, Chrome 91, Safari 14.1

j1elo committed:
README: full review, add section Implementation Notes

j1elo committed:
README: small additions

j1elo committed:
Bump to Typescript 4 and mediasoup 3.8

j1elo committed:
SdpUtils: rename MsSdpUtils to MsSdpCommonUtils

Follow the original naming more closely

j1elo committed:
SdpUtils: use unifiedPlanUtils instead of plainRtpUtils

The way plainRtpUtils parses SDP to obtain encodings information (just a
single, main SSRC of the media) doesn't work for WebRTC, where there is
an SSRC group containing both a main SSRC and a secondary SSRC of the
RTX/retransmissions flow.

Instead, using Unified Plan implementation, we are able to correctly
parse the RTX SSRC and this will later be used by mediasoup to identify
and handle RTX packets.

j1elo committed:
Bump to mediasoup 3.9

j1elo committed:
Fix TypeScript version to Major.Minor instead of just Major

https://github.com/microsoft/TypeScript/issues/14116#issuecomment-280410804

TypeScript never claimed to follow semantic versioning, in the sense that
breaking changes imply major versions.

TypeScript, however, promises no breaking changes after a stable release.
so no breaking changes between 2.1.5 and 2.1.6, 2.1.*.

My recommendation is fix your version of typescript to major.minor instead
of just major. e.g. ^2.1 and not ^2

j1elo committed:
Bump to Node 16

j1elo committed:
processOffer: fix negotiation of producer RtpParameters

j1elo committed:
processOffer: throw on error instead of just logging

j1elo committed:
Add .prettierignore to skip formatting some files

j1elo committed:
Bump to Prettier 2.5

Support for Logical Nullish Assignment (??=) was added in Prettier 2.1

j1elo committed:
package.json: change mediasoup peerDependencies to any 3.x

j1elo committed:
processOffer: detect simulcast and set scalabilityMode "L1T3"

j1elo committed:
package: Bump "engines" to NPM 16.0 LTS

j1elo committed:
SdpUtils: Use MediaAttributes type instead of "any" for SDP objects

j1elo committed:
SdpUtils: Complement encodings with both "ssrcs" and "rids"

On Chrome, an SDP Offer will come with either a list of SSRCs or a list
of RIDs, that's why the code was filling encodings from either one or
the other.

However, Firefox does indeed signal the SSRCs even when using RIDs for
simulcast, which means that both types of identifiers should be
processed.

j1elo committed:
Logging: Use JSON.stringify() instead of "%O" for objects

j1elo committed:
SdpUtils: Fix wrong order of caps when parsing SDP Offer

j1elo committed:
SdpUtils: Rename sendParams, recvCaps to producerParams, consumerCaps

These names express the intention of the value with more clarity. The
method "sdpToSendRtpParameters" is really concerned with generating an
RtpParameters object that is appropriate for making a Producer out of an
SDP Offer, so "sdpToProducerRtpParameters" fits better its purpose.

This also is useful to separate our view of "sending" (a remote endpoint
sent an SDP Offer and is about to send data to us) from the view of
"sending" that is assumed by mediasoup-client (we are sending an SDP
Offer and about to start sending data).

j1elo committed:
Remove usages of process.exit; review text of error logs

j1elo committed:
SdpUtils: Use the correct Header Extension IDs from SDP Offers

j1elo committed:
package.json: Add mediasoup-client to peerDependencies

Be explicit on stating that this library will use the application's
version of mediasoup-client.

j1elo committed:
SdpUtils: Use correct Header Extension and Payload Type for SDP Offers

j1elo committed:
Rewrite error messages to follow same structure

j1elo committed:
SdpUtils: Filter generated caps by media kind

No need to include caps for media kinds that are not the current one; it
didn't hurt either, but at least this way the debug logs show up much
clearer.

j1elo committed:
Work around sdp-transform bug #94: "Type inconsistency in payloads"

When the SDP media line contains just 1 PayloadType number, this is
converted into an int instead of the usual string.

This has no impact in our code right now, but for consistency's sake,
it's better that small bugs like this are not allowed to exist. The
upstream library has an open issue, but doesn't look like it will get
fixed any time soon... so apply an easy workaround here.

cruizba committed:
Bump to mediasoup 3.10

cruizba committed:
Update typescript to version 4.9.5

j1elo committed:
deps: Update to mediasoup 3.11.26, mediasoup-client 3.6.84

Same as mediasoup-controller

j1elo committed:
deps: Update to typescript 4.9.5

Same as mediasoup-controller

j1elo committed:
tools/browser-rtpcapabilities: Use JSON5 to pretty-print JS object code

This replaces printing a JSON object (with quoted keys) and instead
prints a proper JavaScript object string (with unquoted keys), thanks to
the `stringify` method provided by JSON5.

j1elo committed:
BrowserRtpCapabilities: Update with latest caps from Firefox, Chrome

j1elo committed:
tools/browser-rtpcapabilities: Fix errata in README

j1elo committed:
SdpUtils: Rename import "MsSdpCommonUtils" to "MsSdpUtils"

Use "MsSdpUtils" which is the name used by apps using the same import.

j1elo committed:
SdpUtils: Add comment to warn about not using multiple audios/videos

j1elo committed:
Improve error handling

j1elo committed:
browser-rtpcapabilities: Show more details; add RtpParameters

j1elo committed:
browser-rtpcapabilities: Fix undefined adapter

j1elo committed:
Prettier.js: Use default config file name '.prettierrc'

j1elo committed:
SdpUtils: Improve handling of RtpCapabilities (validate input)

j1elo committed:
SdpUtils: Debug log proper RtpParameters/RtpCapabilities

j1elo committed:
browser-rtpcapabilities: Log everything needed for LiveKit translation

j1elo committed:
Fix OpenVidu bug: Skip more than 1 media of each kind

OpenVidu E2E tests started to show a behavior of sending SDP Offers with
2 audio medias. This is not compatible with mediasoup-sdp-bridge, and
caused that mediasoup would end up erroring out.

There's no strong reason to require exactly 1 audio, apart from forcing
the client to be exactly right. But it's worth being more flexible here
and give time some headroom to improve without having broken tests.

mcottontensor committed:
Removing reference to our mediasoup prebuilt package since mediasoup provide binaries for their workers now.

mcottontensor committed:
Fixing up mediasoup uses for the latest version

lukehb commented on SFU/sfu_server.js:
```diff
@@ -298,9 +298,6 @@ async function setupPeerDataChannels(peerId) {
     // streamer data consumer (consumes peer data)
     peer.streamerDataConsumer = await streamer.transport.consumeData({ dataProducerId: peer.peerDataProducer.id });
 
-    streamer.transport._sctpStreamIds[nextStreamerSCTPStreamId] = 1;
-    streamer.transport._sctpStreamIds[nextPeerSCTPStreamId] = 1;
-
```
I assume these lines were significant? Are they correctly set to the appropriate values in the new version?

mcottontensor replied:
So that property has been locked down by mediasoup so we cant get to it without hacky bs. I went looking through the code to see if the behaviour was auto handled or something now but couldn't nail it down completely. So I just removed it and tried the SFU with multiple connections. I was able to control the stream on both connections and I noticed the output showing that the stream id values were increasing. So it appears to not be needed.

lukehb replied:
Sounds good them

lukehb reviewed (commented):
Just want to sanity check two lines that were removed.

mcottontensor committed:
Rather than hosting the mediasoup-sdp-bridge package here as raw JS. I've created a new repo that hosts the TS project and publishes the actual package. This way it now has type information and history.

mcottontensor committed:
Add 'Extras/mediasoup-sdp-bridge/' from commit '090eb4482937ad6ebe0f46b599e5354603aa043d'

git-subtree-dir: Extras/mediasoup-sdp-bridge
git-subtree-mainline: 0bd8db09839791e762adec83db91831afcaf2b1a
git-subtree-split: 090eb4482937ad6ebe0f46b599e5354603aa043d

mcottontensor committed:
Updating package.json to our forked details. Updated mediasoup to latest.

mcottontensor committed:
Adding datachannel support

mcottontensor committed:
Adding scalability mode support

mcottontensor committed:
Adding probator change

mcottontensor committed:
Removed sendonly check

mcottontensor committed:
Adding notes to explain how to manage this project.

mcottontensor committed:
Updating workspace to include the new mediasoup-sdp-bridge package

mcottontensor committed:
Update NOTES.md

mcottontensor committed:
Cleaning up some of the datachannel behaviour to make it a little less hacky.

mcottontensor committed:
Updating to match the changes in the sdp-bridge. The data channel handling is a little less hacky now. Also fixed up some old naming on some missed function calls.
